src/rust/receiver-mock/src/router/otlp.rs

```diff
@@ -0,0 +1,296 @@
+) -> impl Responder {
+    // Don't fail when we can't read remote address.
+    // Default to localhost and just ingest what was sent.
+    let localhost: std::net::SocketAddr = std::net::SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 0);
+    let remote_address = req.peer_addr().unwrap_or(localhost).ip();
+
+    let headers = req.headers();
+    let empty_header = HeaderValue::from_str("").unwrap();
+    let content_type = headers
+        .get("content-type")
+        .unwrap_or(&empty_header)
+        .to_str()
+        .unwrap();
+
+    if let Some(response) = try_dropping_data(&opts, content_type) {
+        return response;
+    }
+
+    match content_type {
+        OTLP_PROTOBUF_FORMAT_CONTENT_TYPE => {
+            let log_data: logsv1::LogsData = match logsv1::LogsData::decode(&mut Cursor::new(body)) {
+                Ok(data) => data,
+                Err(_) => return HttpResponse::BadRequest().body("Unable to parse body"),
+            };
+            for resource_logs in log_data.resource_logs {
+                let metadata = get_otlp_metadata_from_logs(&resource_logs);
+                let lines = get_otlp_lines_from_logs(&resource_logs);
+
+                app_state.add_log_lines(
+                    lines.iter().map(|x| x.as_str()),
+                    metadata,
+                    remote_address,
+                    &opts,
+                );
+
+                if opts.print.logs {
+                    for line in lines {
+                        println!("log => {}", line);
+                    }
+                }
+            }
+        }
+        &_ => {
+            return HttpResponse::build(StatusCode::BAD_REQUEST).json(ReceiverError {
+                id: String::from(DUMMY_ERROR_ID),
+                errors: vec![ReceiverErrorErrorsField {
+                    code: String::from("header:invalid"),
+                    message: format!("Invalid Content-Type header: {}", content_type),
+                }],
+            })
+        }
+    }
+
+    HttpResponse::Ok().body("")
+}
+
+fn get_otlp_metadata_from_logs(resource_logs: &logsv1::ResourceLogs) -> Metadata {
+    match &resource_logs.resource {
+        Some(resource) => HashMap::from_iter(resource.attributes.iter().map(|kv| {
+            (
+                kv.key.clone(),
+                match &kv.value {
+                    Some(value) => otlp_anyvalue_to_string(value),
+                    None => String::new(),
+                },
+            )
+        })),
+        None => Metadata::new(),
+    }
+}
+
+fn get_otlp_lines_from_logs(resource_logs: &logsv1::ResourceLogs) -> Vec<String> {
+    resource_logs
+        .instrumentation_library_logs
+        .iter()
+        .map(|ill| {
+            ill.log_records.iter().map(|log_record| match &log_record.body {
+                Some(body) => otlp_anyvalue_to_string(&body),
+                None => String::new(),
+            })
+        })
+        .flatten()
+        .collect()
```

Comment:
I think this would be clearer in a different order:
```suggestion
        .map(|ill| ill.log_records.iter())
        .flatten()
        .map(|log_record| match &log_record.body {
            Some(body) => otlp_anyvalue_to_string(&body),
            None => String::new(),
        })
        .collect()
```
Less nesting this way.